### DISCUSSION TITLE
Running OneTimeSetup and OneTimeTearDown only once per execution

### DISCUSSION THREAD

**Discussion by ayankumaar** [nunit/nunit#3785] [Q&A] [answered]
I am running a dotnet test command inside a loop. I am running it like this:

**dotnet test --filter TestCategory="Smoke"  "C:\\MyTest\\WorkSpace\\MyTest\\bin\\Debug\\netcoreapp3.1\\MyTest.exe"**

**The above mentioned command gets executed 5 times inside a loop**. But, I want few things to get executed only once. So, I have written those inside OneTimeSetup and OneTimeTearDown. But, OneTimeSetup and OneTimeTearDown are also getting executed each and every time with the dotnet test command.

Is there any way to execute OneTimeSetup and OneTimeTearDown only once for the above mentioned scenario ? Or any other way to achieve that ?

**Comment by mikkelbu** [comment]
If you are running `dotnet test` inside a loop,, then you are starting the test processes several times, so the logic inside NUnit cannot help you as no state is kept between process invocations. So you need to code this yourself. One solution could be to create a file to act as a lock, then check if this file exists (e.g. inside the OneTimeSetup) - and if it don't run the initialization and if it does then do nothing. After the loop you remove the file again.

Please note that this simple solution will not work if you plan to run several instances of the loop in parallel, then you need to do something smarter.

**Reply by ayankumaar** [reply]
@mikkelbu Thanks for your reply. Is there any way to run a script 5 times using a single command ? Something like this

dotnet test --filter TestCategory="Smoke" --iterations 5

**Reply by mikkelbu** [reply]
I don't think so. Also `dotnet test` is not NUnit code, but .NET CLI calling the NUnit adapter, so even if there where such an option it still won't work, as the iteration is done outside NUnit.

Why do you want to run the tests 5 times? (are they flaky)
And why do you want to only do the initialization once? (it is expensive or is there another reason)

**Reply by ayankumaar** [reply]
@mikkelbu I need to perform an action only once and a sequence of actions after it for 5 times and after that another action only once. You can think of it like this:
 action1 - 1 time, 
action2 - 5 times, 
action3 - 1 time.

These needs to happen sequentially

What is the best way to achieve this in NUnit ?

I thought of this way

[OneTimeSetup]
 public void Setup(){
   //action1
}

[Test]
[Category="Smoke"]
public void PerfAction(){
   // action2
}
 

[OneTimeTearDown]
public void TearDown(){
      // action3
}
 
But, my method doesn't seem to be working. So, I am looking for a way to perform iterative testing in NUnit. Is there any way to do that ?

**Reply by mikkelbu** [reply]
One simple way would be to make PerfAction take an argument and make 5 different test cases. So something like this
```csharp
....
[TestCase(1)]
[TestCase(2)]
[TestCase(3)]
[TestCase(4)]
[TestCase(5)]
[Category="Smoke"]
public void PerfAction(int dummyValue){
// action2
}
...
```
But if you want it to be more general then you can use a TestCaseSource that can supply the individual testcases. Whether it is the best way very much depends on why you want to run it 5 times and what the other steps are.

**Reply by ayankumaar** [reply]
@mikkelbu In my case the number of times I will execute PerfAction is not fixed all the time. Sometimes it may be 3 or sometimes may be more or less than that. Basically, it can be anything. So, in your approach I will have to configure my script each and every time I run it. Rather than that I will prefer it to be specified from a file, pick the value and execute based on that value passed. That is the approach I am looking at. So, is there any way to implement it in that way ? Also, any argument that comes with the dotnet test command ?

To make your understanding simpler, I am actually trying to perform iterative testing (data driven testing) in NUnit. So, what is the way to achieve that ?

**Reply by mikkelbu** [reply, marked as answer]
@ayankumaar That was also why I wrote
> But if you want it to be more general then you can use a TestCaseSource that can supply the individual testcases. 

This TestCaseSource can then provide the number of testcases that you desire. So the method can read the data from a file and then create the number of testcases specified in the file.

I don't think `dotnet test` supports passing parameters to the underlying test adapter, but that can have changed since - see e.g. https://github.com/nunit/nunit-console/issues/404#issuecomment-381451679. An alternative approach is to use environment variables to pass the information.
